### ISSUE TITLE
Can't connect an error (size_out=1) to a learning rule object with size_in > 1

### ISSUE THREAD

**Issue by s72sue** [nengo/nengo#772]
There seems to be no way to connect an error (with size_out=1) to a learning rule object with size_in=3.  It seems like you need to specify a (3x1) transform:
![image](https://cloud.githubusercontent.com/assets/12284191/8555525/a37d9eda-24be-11e5-803b-8dcdaa945e93.png)
Or a function:
![test2](https://cloud.githubusercontent.com/assets/12284191/8556502/632a46d8-24c5-11e5-83b9-223c231ca245.png)

However, both of the above approaches give a run time erorr in PES:
![image](https://cloud.githubusercontent.com/assets/12284191/8555466/531cbd04-24be-11e5-9802-c8af5667d21c.png)

**Comment by hunse** [comment]
Can you post a whole minimal example?

**Comment by s72sue** [comment]
I have created a simple example shown below  ( you can also download the iPython notebook from here if you want to play with it: https://drive.google.com/file/d/0B90f2TExXnVlU2xxUkFhc0VHWXc/view?usp=sharing ). Also note the change in the title (basically it doesn't work if the learning rule has size_in>1).

![test1](https://cloud.githubusercontent.com/assets/12284191/8571924/8f07d59e-2557-11e5-9c54-99c2f07bfe38.png)

**Comment by arvoelke** [comment]
If you set `weights=False` then it works. I think the issue is that when you use a weight solver, you no longer have access to the vector space (i.e. the decoders), and so PES updates the entire weight matrix instead. Still not quite sure why I couldn't get it to work. Try looking at the example here: https://github.com/nengo/nengo/blob/master/nengo/tests/test_learning_rules.py#L77

Specifically the unit test `test_pes_weight_solver`. Switching to Eric's branch `monolithic-transform` may also help in some respects. Also going to the lines in `nengo` where the exceptions are being thrown can be helpful.

**Comment by drasmuss** [comment]
I don't think the PES rule is set up to work on transformation matrices right now, it only works on decoders (`PES.error_type == 'decoders'`).  And it looks like the `LearningRule.size_in` is set up to only work with `decoders` as well.  That's what setting `weights=True` does, it (in theory) tells the PES rule that it should be learning a transformation matrix, not decoders.  But while the PES rule itself is set up to support that, it doesn't look like the general learning rule code is.

**Comment by arvoelke** [comment]
Prior to the `monolithic-transform` branch #729, the error connection for PES was to be applied before the transform (after the decoders). But with that change, the transform goes into the decoders, and so the error connection is now with respect to `size_mid` (the dimensionality of the transformed output). It should be possible to get this working.

**Comment by drasmuss** [comment]
PS here is a minimal example:
```python
import nengo
import numpy as np

with nengo.Network() as net:
    err = nengo.Node(output=[0, 0])
    ens1 = nengo.Ensemble(20, 2)
    ens2 = nengo.Ensemble(20, 1)

    conn = nengo.Connection(ens1, ens2,
                            learning_rule_type={"pes": nengo.PES(1e-6)},
                            transform=np.random.randn(1, 2),
                            solver=nengo.solvers.LstsqL2(weights=False))
    nengo.Connection(err, conn.learning_rule["pes"])

sim = nengo.Simulator(net)
sim.run(1.0)
```

**Comment by drasmuss** [comment]
Note that if you did want to apply the learning rule to the connection weights, not the decoders, the error signal should be 1D, not 2D (but that doesn't work right now either because of the `size_in` things).

**Comment by s72sue** [comment]
Yeah I do want to apply the learning rule to the connections weights and
not to the decoders. The error signal that I am using (and also in the
example I provided is 1D), but only the size_in of the learning rule object
is > 1.  I have implemented this sort of learning before using the older
learning interface (with modulatory connections) and it worked fine.

**Comment by drasmuss** [comment]
Sorry, I meant the dimensionality of the error signal actually driving the learning rule (i.e. the size_out of the error connection).

**Comment by s72sue** [comment]
Do you mean size_out of 'conn_1' in the example above?  If yes, then it is has a size_out=1 and size_in=2  :) .  If you meant something else, I am sorry I din't understand it  :(

**Comment by drasmuss** [comment]
The error connection, so `error_conn1` in the example.

**Comment by arvoelke** [comment]
> I have implemented this sort of learning before using the older
learning interface (with modulatory connections) and it worked fine.

Once you get it working (should be possible with the unit test I pointed you to, and the fact about PES error going before/after the transform depending if you're before/after #729), would be good to add a unit test to that same file in order to ensure that it doesn't change again. Likely an oversight in coverage on our part.

**Comment by s72sue** [comment]
@drasmuss  I see what you mean, thanks!
@arvoelke  I havn't taken a look at the unit test yet, but I will take a look. Thanks for pointing to it.

**Comment by s72sue** [comment]
fixed by  #729

**Comment by tbekolay** [comment]
I forgot about this! I'll add a quick test case so that we can close this issue.
